mavlink/message.go

```diff
@@ -102,53 +103,91 @@ func newPacketFromBytes(b []byte) (*Packet, int) {
 // a message they're interested in, and convert it to the
 // corresponding type via Message.FromPacket()
 func (dec *Decoder) Decode() (*Packet, error) {
-
-	// discard bytes until our start byte
 	for {
-		c, err := dec.br.ReadByte()
-		if err != nil {
-			return nil, err
+		startFoundInBuffer := false
+		// discard bytes in buffer before start byte
+		for i, b := range dec.buffer {
+			if b == startByte {
+				dec.buffer = dec.buffer[i:]
+				startFoundInBuffer = true
+				break
+			}
 		}
-		if c == startByte {
-			break
+
+		// if start not found, read until we see start byte
+		if !startFoundInBuffer {
+			for {
+				c, err := dec.br.ReadByte()
+				if err != nil {
+					return nil, err
+				}
+				if c == startByte {
+					dec.buffer = append(dec.buffer, c)
+					break
+				}
+			}
 		}
-	}
 
-	// hdr contains LENGTH, SEQ, SYSID, COMPID, MSGID
-	hdr := make([]byte, 5)
-	if _, err := io.ReadFull(dec.br, hdr); err != nil {
-		return nil, err
-	}
+		if len(dec.buffer) < 2 {
+			// read length byte
+			bytesRead := make([]byte, 1)
```

Comment:
I ran the following on an edison:
```
func testSliceAlloc() {
	oneByteTime := int64(0)
	multipleByteTime := int64(0)
	numIterations := int64(1000000)

	for i := int64(0); i < numIterations; i++ {
		t := time.Now().UnixNano()
		x := make([]byte, 1)
		oneByteTime += time.Now().UnixNano() - t
		x[0] = 0 // use variable to appease compiler

		t = time.Now().UnixNano()
		x = make([]byte, 263)
		multipleByteTime += time.Now().UnixNano() - t
		x[0] = 0
	}

	fmt.Printf("%d average time ns for make([]byte, 1) with %d iterations\n", oneByteTime/numIterations, numIterations)
	fmt.Printf("%d average time ns for make([]byte, 263) with %d iterations\n", multipleByteTime/numIterations, numIterations)
}
```
and got the following output:
```
2050 average time ns for make([]byte, 1) with 1000000 iterations
2596 average time ns for make([]byte, 263) with 1000000 iterations
```
Let's just round up and say these allocations take 3μs on average.

If the autopilot is connected via RS-232 serial using 115200 for the baud rate, and no parity bit, the time to transfer 263 bytes will be at least `263 * 10 / 115200` which is about 22.8ms, or over 7500 times slower. (263 bytes is used as an example because that's the largest size a mavlink packet can be).

In other words, the memory allocation here is orders of magnitude faster than the receiving of bytes from the autopilot.

Running the same code above using `runtime/pprof` to get a memory profile, only 257 Mallocs are made, i.e., the runtime for the most part is reusing memory it has already allocated.

Anyway, it'd probably be best to use a single circular buffer that only gets allocated once, but in this case, I don't think it matters.